# Tutorial: How to set up a geospatial collection of counties for a U.S. State with ckanext-gztr using Python, Apache SedonaDB, and TIGER/Line Shapefiles

We'll demonstrate how to catalog the counties for a U.S. state from a [TIGER/Line Shapefile](https://www.census.gov/geographies/mapping-files/time-series/geo/tiger-line-file.html) on a CKAN instance with ckanext-gztr installed. The steps should be similar for other shapefiles provided by the data source which you can modify.

In this tutorial we will go over how to set up a STAC collection for your CKAN instance that has ckanext-gztr installed. In particular we will download HUC08 Sub Basins data for your intended region from the Geoconnex reference collections API.

Learn more:

- ckanext-gztr: [gztr.dathere.com](https://gztr.dathere.com)

In this tutorial we'll attempt to:

1. Retrieve the national counties Shapefile from the U.S. Census Bureau.
2. Filter this data to a single state's counties (e.g. New York).
3. Filter the metadata to only what's necessary for ckanext-gztr.
4. Upload a GeoParquet of the data to our CKAN instance's storage for ckanext-gztr.
5. Update the `catalog.json` and `collections.json` files to link to the new file so that it is available in the STAC API.

## Setup your libraries and SedonaDB connection

First we'll install external Python libraries:

- [Apache SedonaDB](https://sedona.apache.org/sedonadb/latest/) - For setting up a geospatial [DataFrame](https://sedona.apache.org/sedonadb/latest/reference/python/#sedonadb.dataframe.DataFrame)
- [lonboard](https://developmentseed.org/lonboard/latest) - For visualizing our DataFrame in an interactive map
- [jupyterlab](https://pypi.org/project/jupyterlab/) - Required for lonboard to work properly in a Jupyter Lab (based on [this issue](https://github.com/developmentseed/lonboard/issues/397))
- [ckanapi](https://github.com/ckan/ckanapi) - For interacting with your CKAN instance's API (to upload the data and modify your STAC Catalog)
- [requests](https://requests.readthedocs.io/en/latest/) - For retrieving your STAC `catalog.json` and `collections.json` data when we update them

We'll be using [Apache SedonaDB](https://sedona.apache.org/sedonadb/latest/) to make a dataframe we can use to import, transform, and export the geospatial data as required along with providing access to spatial SQL.

Start by installing it with a package manager such as `uv` or `pip`.

In [1]:
# The dependencies should already be installed if you're using mybinder.org
#   If not then uncomment and run one of the following commands.
#   Then you'll need to refresh your browser to use the interactive lonboard visualizations.
# !pip install "apache-sedona[db]" lonboard jupyterlab ckanapi requests
# or if you have uv installed
# !uv pip install "apache-sedona[db]" lonboard jupyterlab ckanapi requests

Next we'll import the package and set up the query interface.

In [2]:
import requests
import sedona.db
from ckanapi import RemoteCKAN
from lonboard import viz
from pprint import pprint

In [3]:
sd = sedona.db.connect()

## Find the national counties Shapefile and download it

The U.S. Census Bureau provides a national Shapefile which includes all U.S. counties along with associated metadata for each.


## Read the TIGER/Line Shapefile into a SedonaDB dataframe

SedonaDB has a convenient [`read_pyogrio`](https://sedona.apache.org/sedonadb/latest/reference/python/?h=pyogrio#sedonadb.context.SedonaContext.read_pyogrio) function to read spatial formats using GDAL/OGR. This includes support for GeoJSON, GeoPackage, FlatGeoBuf, and for our use case we'll use it for reading a Shapefile by providing the URL into a SedonaDB [`DataFrame`](https://sedona.apache.org/sedonadb/latest/reference/python/?h=pyogrio#sedonadb.dataframe.DataFrame).

In [4]:
df = sd.read_pyogrio("https://www2.census.gov/geo/tiger/TIGER2025/COUNTY/tl_2025_us_county.zip")

## Filter for the state of New York's counties

Let's explore the columns available in our dataframe with `df.columns`.

In [5]:
for column in df.columns:
    print(column)

STATEFP
COUNTYFP
COUNTYNS
GEOID
GEOIDFQ
NAME
NAMELSAD
LSAD
CLASSFP
MTFCC
CSAFP
CBSAFP
METDIVFP
FUNCSTAT
ALAND
AWATER
INTPTLAT
INTPTLON
wkb_geometry


For this tutorial we'll assume we want to only add counties from the state of New York for our CKAN instance. We can filter the `STATEFP` column for the Federal Information Processing Standards (FIPS) code of New York which is `36`.

In [6]:
df = df.filter(df["STATEFP"] == 36)

Just to confirm we have 62 counties of New York at the time this notebook was ran:

In [7]:
df.count()

62

It is also important to check the Coordinate Reference System (CRS) used for the geometry of the shapefile. We can view the geometry used by running `df.schema` to look at the type of the `wkb_geometry` column.

In [8]:
df.schema

SedonaSchema with 19 fields:
  STATEFP: utf8<Utf8>
  COUNTYFP: utf8<Utf8>
  COUNTYNS: utf8<Utf8>
  GEOID: utf8<Utf8>
  GEOIDFQ: utf8<Utf8>
  NAME: utf8<Utf8>
  NAMELSAD: utf8<Utf8>
  LSAD: utf8<Utf8>
  CLASSFP: utf8<Utf8>
  MTFCC: utf8<Utf8>
  CSAFP: utf8<Utf8>
  CBSAFP: utf8<Utf8>
  METDIVFP: utf8<Utf8>
  FUNCSTAT: utf8<Utf8>
  ALAND: int64<Int64>
  AWATER: int64<Int64>
  INTPTLAT: utf8<Utf8>
  INTPTLON: utf8<Utf8>
  wkb_geometry: geometry<Wkb(epsg:4269)>

Notice how the geometry is using the `epsg:4269` CRS, meanwhile currently ckanext-gztr expects the CRS to be in `EPSG:4326`.

Since we need to rename some columns and transform the CRS, we'll assign a view to our data so that we can run a spatial SQL query and refer to the view's name directly. In this case we'll use the [`to_view`](https://sedona.apache.org/sedonadb/latest/reference/python/?h=pyogrio#sedonadb.dataframe.DataFrame.to_view) function and assign the name `counties`.

In [9]:
df.to_view("counties")

Now we'll want to filter our dataframe to just the columns that we need for ckanext-gztr:

- An `id` column which is unique across all features in our geospatial collection. We'll use each county's `COUNTYFP` value.
- A `title` column which is a human-readable name for our feature. In this case it's the county's name from the `NAME` column.
- A `geometry` column which is the spatial representation of our county. In this case we'll rename the `wkb_geometry` column to `geometry` and also change its projection from `epsg:4269` to `epsg:4326` using the [`ST_Transform`](https://sedona.apache.org/sedonadb/latest/reference/sql/#st_transform) spatial SQL function.

In [10]:
df = sd.sql("""
SELECT
    "COUNTYFP" AS id,
    "NAME" AS title,
    ST_Transform("wkb_geometry", 'EPSG:4326') AS geometry
FROM counties
""")
df.show()

┌──────┬────────────┬──────────────────────────────────────────────────────────────────────────────┐
│  id  ┆    title   ┆                                   geometry                                   │
│ utf8 ┆    utf8    ┆                                   geometry                                   │
╞══════╪════════════╪══════════════════════════════════════════════════════════════════════════════╡
│ 123  ┆ Yates      ┆ POLYGON((-77.35735 42.669193,-77.357337 42.669301,-77.357333 42.669337,-77.… │
├╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 039  ┆ Greene     ┆ POLYGON((-73.794323 42.376822,-73.794316 42.376188,-73.794327 42.375084,-73… │
├╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 043  ┆ Herkimer   ┆ POLYGON((-75.167806 44.095939,-75.15712 44.091302,-75.155738 44.090702,-75.… │
├╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌

We can visualize the counties on an interactive map using the `viz` function from the `lonboard` Python library (by first providing a GeoPandas dataframe based on our Apache SedonaDB dataframe using the [`to_pandas()`](https://sedona.apache.org/sedonadb/latest/reference/python/#sedonadb.dataframe.DataFrame.to_pandas) function).

In [11]:
viz(df.to_pandas())

Now our geospatial collection is formatted as required for ckanext-gztr. We'll save our geospatial collection to a GeoParquet file using the [`to_parquet`](https://sedona.apache.org/sedonadb/latest/reference/python/?h=pyogrio#sedonadb.dataframe.DataFrame.to_parquet) function.

In [12]:
df.to_parquet("ny_counties.parquet")

We'll want to upload the GeoParquet file to our CKAN instance's storage used by ckanext-gztr. We'll use the `ckanapi` library to help with this.

Now we'll use the `RemoteCKAN` class to provide the URL to our CKAN instance, a sysadmin `apikey`, and then run the [`file_create`](https://docs.ckan.org/en/latest/api/index.html#ckan.logic.action.file.file_create) Action API endpoint to upload our `ny_counties.parquet` file to the `gztr` storage. Then we'll print the response because we'll need the stem of the `location` value from the response for updating `catalog.json` and `collections.json`.

In [13]:
from ckanapi import RemoteCKAN

CKAN_INSTANCE_URL = "http://localhost:5000"
CKAN_API_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJqdGkiOiJjeXhpc1g2SHlOZjNkQzlPZWZXekZNbHE2Z0Nham5vS3ZTaktFTnZ3ZnNnIiwiaWF0IjoxNzg3MDUzMTcwfQ.EUftyGDkEx312TA_5S-FxcIaYDwIgiNSyASEi-juoVk"

ckan = RemoteCKAN(CKAN_INSTANCE_URL, apikey=CKAN_API_KEY)
response = ckan.action.file_create(storage="gztr", upload=open("ny_counties.parquet", "rb"))
print(response)

{'name': 'ny_counties.parquet', 'location': 'ny_counties.parquet', 'storage': 'gztr', 'content_type': 'application/octet-stream', 'size': 1183112, 'hash': 'd1e848bcc943b676d7e52cd5c5679431', 'algorithm': 'md5', 'created': '2026-09-01T05:29:40.524196+00:00', 'storage_data': {}, 'id': '84860de8-51ee-44c7-9253-03fdf7314b50', 'owner_type': 'user', 'owner_id': '255d924c-f5fc-4dc1-9204-1e9bfa899219', 'pinned': False}


We've now uploaded our GeoParquet file to the `gztr` storage. Note the `location` value's stem is `ny_counties`. We'll need to update `catalog.json` and `collections.json` to add this new geospatial collection as a STAC collection.

We will:

- Define the `location_stem` as `ny_counties` (based on the `location` value returned from the `file_create` response in the cell above.
- Calculate the spatial extent (the bounding box) of all New York counties using the `total_bounds` of a GeoPandas dataframe.
- Calculate a temporal start datetime timestamp value based on the current time in UTC (you could change the temporal interval start and end time to a different time if that makes more sense, for example since the original shapefile is for the Gregorian year 2025, you could label the start and end temporal extent values as timestamps based on that datetime range).
- Download the `catalog.json` and `collections.json` files using the `requests` library and load the output into a Python dictionary. Then append a new collection to the `links` list from the STAC catalog and a new collection to the `collections.json` list.
- Save the STAC catalog and collections as files, delete the current remote STAC files, and then create new files based on the new local files.

In [14]:
location_stem = "ny_counties"

In [15]:
# Calculate the spatial extent (extent.spatial.bbox)
bbox = df.to_pandas().total_bounds.tolist()
bbox

[-79.76259, 40.476578, -71.777491, 45.015865]

In [16]:
# For temporal interval value (we only calculate the start timestamp for now)
import datetime
# Get the current timestamp in accordance with RFC 3339
start_time = datetime.datetime.now(datetime.UTC).strftime("%Y-%m-%dT%H:%M:%SZ")
start_time

'2026-09-01T05:30:05Z'

In [17]:
import requests
from pprint import pprint

In [18]:
catalog = requests.get(f"{CKAN_INSTANCE_URL}/gztr/stac").json()
catalog["links"].append({
    "href": f"{CKAN_INSTANCE_URL}/gztr/stac/collections/{location_stem}",
    "rel": "child",
    "title": "New York Counties",
    "type": "application/json"
})
pprint(catalog)

{'description': 'Geospatial collections and features used for dataset '
                'publishing and search by the New Mexico Water Data Catalog, '
                'organized through the SpatioTemporal Asset Catalogs (STAC) '
                'specification.',
 'id': 'nmwdc',
 'links': [{'href': 'http://localhost:5000/gztr/stac',
            'rel': 'self',
            'type': 'application/json'},
           {'href': 'http://localhost:5000/gztr/stac',
            'rel': 'root',
            'type': 'application/json'},
           {'href': 'http://localhost:5000/gztr/stac/collections',
            'rel': 'collections',
            'type': 'application/json'},
           {'href': 'http://localhost:5000/gztr/stac/collections/public_water_systems',
            'rel': 'child',
            'title': 'Public Water Systems',
            'type': 'application/json'},
           {'href': 'http://localhost:5000/gztr/stac/collections/nm_huc08_sub_basins',
            'rel': 'child',
            'tit

In [19]:
collections = requests.get(f"{CKAN_INSTANCE_URL}/gztr/stac/collections").json()

COLLECTION_TITLE = "New York Counties"
COLLECTION_DESCRIPTION = "Counties in the state of New York within the United States. Data provided by TIGER/Line Shapefiles."

collections.append({
  "stac_version": "1.1.0",
  "title": COLLECTION_TITLE,
  "type": "Collection",
  "description": COLLECTION_DESCRIPTION,
  "extent": {
    "spatial": {
      "bbox": [bbox]
    },
    "temporal": {
      "interval": [start_time,None]
    }
  },
  "id": location_stem,
  "license": "proprietary",
  "links": [
    {
      "href": f"{CKAN_INSTANCE_URL}/gztr/stac/collections/{location_stem}",
      "rel": "self",
      "type": "application/json"
    },
    {
      "href": f"{CKAN_INSTANCE_URL}/gztr/stac/collections",
      "rel": "parent",
      "type": "application/json"
    },
    {
      "href": f"{CKAN_INSTANCE_URL}/gztr/stac",
      "rel": "root",
      "type": "application/json"
    },
    {
      "href": f"{CKAN_INSTANCE_URL}/gztr/stac/collections/{location_stem}/items",
      "rel": "items",
      "type": "application/geo+json"
    }
  ],
  "providers": [
    {
      "description": "United States government agency providing demographic and economic data.",
      "name": "U.S. Census Bureau",
      "roles": [
        "producer",
        "host"
      ],
      "url": "https://census.gov"
    }
  ]
})
print(collections)

[{'description': 'Public Water Systems as defined by the U.S. Environmental Protection Agency (EPA) which intersect with New Mexico.', 'extent': {'spatial': {'bbox': [[-180.0, -90.0, 180.0, 90.0]]}, 'temporal': {'interval': [[None, None]]}}, 'id': 'public_water_systems', 'license': 'proprietary', 'links': [{'href': 'http://localhost:5000/gztr/stac/collections/public_water_systems', 'rel': 'self', 'type': 'application/json'}, {'href': 'http://localhost:5000/gztr/stac/collections', 'rel': 'parent', 'type': 'application/json'}, {'href': 'http://localhost:5000/gztr/stac', 'rel': 'root', 'type': 'application/json'}, {'href': 'http://localhost:5000/gztr/stac/collections/public_water_systems/items', 'rel': 'items', 'type': 'application/geo+json'}], 'providers': [{'description': 'Independent agency of the United States government focused on environmental protection concerns.', 'name': 'U.S. Environmental Protection Agency (EPA)', 'roles': ['producer'], 'url': 'https://epa.gov'}, {'description'

In [20]:
import json

with open("catalog.json", "w") as f:
    json.dump(catalog, f)
with open("collections.json", "w") as f:
    json.dump(collections, f)

In [21]:
storage_files = ckan.action.file_owner_scan(storage="gztr")["results"]

In [22]:
old_catalog_file_id = [file for file in storage_files if file["location"] == "catalog.json"][0]["id"]
ckan.action.file_delete(storage="gztr", id=old_catalog_file_id)
new_catalog_create_response = ckan.action.file_create(storage="gztr", upload=open("catalog.json", "rb"))

In [23]:
old_collections_file_id = [file for file in storage_files if file["location"] == "collections.json"][0]["id"]
ckan.action.file_delete(storage="gztr", id=old_collections_file_id)
new_collections_create_response = ckan.action.file_create(storage="gztr", upload=open("collections.json", "rb"))

In [24]:
new_catalog = requests.get(f"{CKAN_INSTANCE_URL}/gztr/stac").json()
new_catalog

{'description': 'Geospatial collections and features used for dataset publishing and search by the New Mexico Water Data Catalog, organized through the SpatioTemporal Asset Catalogs (STAC) specification.',
 'id': 'nmwdc',
 'links': [{'href': 'http://localhost:5000/gztr/stac',
   'rel': 'self',
   'type': 'application/json'},
  {'href': 'http://localhost:5000/gztr/stac',
   'rel': 'root',
   'type': 'application/json'},
  {'href': 'http://localhost:5000/gztr/stac/collections',
   'rel': 'collections',
   'type': 'application/json'},
  {'href': 'http://localhost:5000/gztr/stac/collections/public_water_systems',
   'rel': 'child',
   'title': 'Public Water Systems',
   'type': 'application/json'},
  {'href': 'http://localhost:5000/gztr/stac/collections/nm_huc08_sub_basins',
   'rel': 'child',
   'title': 'HUC08 Sub Basins',
   'type': 'application/json'},
  {'href': 'http://localhost:5000/gztr/stac/collections/ny_counties',
   'rel': 'child',
   'title': 'New York Counties',
   'type': 'a

Finally delete the existing `catalog.json` and `collections.json` files and replace them with the updated files:

The New York counties geospatial collection should now be available through the STAC API.